core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1653,29 +1653,32 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     * scala> Stream(1, 2, 3, 4, 5).intersperse(0).toList
     * res0: List[Int] = List(1, 0, 2, 0, 3, 0, 4, 0, 5)
     * }}}
-    */
-  def intersperse[O2 >: O](separator: O2): Stream[F, O2] =
-    this.pull.echo1.flatMap {
-      case None => Pull.done
-      case Some(s) =>
-        s.repeatPull {
-          _.uncons.flatMap {
-            case None => Pull.pure(None)
-            case Some((hd, tl)) =>
-              val interspersed = {
-                val bldr = Vector.newBuilder[O2]
-                bldr.sizeHint(hd.size * 2)
-                hd.foreach { o =>
-                  bldr += separator
-                  bldr += o
-                }
-                Chunk.vector(bldr.result())
-              }
-              Pull.output(interspersed) >> Pull.pure(Some(tl))
-          }
-        }.pull
-          .echo
+    *
+    * This method preserves the Chunking structure of `this` stream.
+    */
+  def intersperse[O2 >: O](separator: O2): Stream[F, O2] = {
+    def doChunk(hd: Chunk[O], isFirst: Boolean): Chunk[O2] = {
+      val bldr = Vector.newBuilder[O2]
```

Comment:
Not related to the PR but vector-based chunks are generally slow unless we know the chunks are big enough to overcome the constant factors. I think we're better off with a mutable buffer here like is done inside various chunk combinators.